### PR TITLE
Add method to list all objects matching a given prefix

### PR DESCRIPTION
**Pull request by hashhar**
All existing methods in TrinoS3Filesystem list objects rooted at the
provided path. This makes listing inefficient for flat structures where
we require multiple API calls to S3 and end up filtering on the client
side. A new method is added which lists all objects whose absolute path
matches a given prefix. This makes searching for all objects matching a
provided prefix efficient. All other methods retain their existing
behavior in this commit.

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 1]
```diff
@@ -605,16 +612,32 @@ public boolean isFilesOnly()
...
+
+    /**
+     * List all objects whose absolute path matches the provided prefix.
+     */
+    private Iterator<LocatedFileStatus> listPrefix(Path prefix, OptionalInt initialMaxKeys, ListingMode mode)
+    {
+        return listPrefix(keyFromPath(prefix), initialMaxKeys, mode);
+    }
+
+    private Iterator<LocatedFileStatus> listPrefix(String prefix, OptionalInt initialMaxKeys, ListingMode mode)
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:636]: It looks like this is only used by the above method, so we should just inline it.

**Review by electrum — APPROVED** [review]

**Comment by hashhar** [comment]
AC as fixup @electrum. PTAL - will squash if it looks okay.

**Comment by findepi** [comment]
@hashhar the fixup is perfect, since this is literally what https://github.com/trinodb/trino/pull/7474#discussion_r608102356 asked for (inlining)

please squash

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:631]: This one is used in two places, but the above one (`listPrefix(Path prefix, ..`) is a good candidate for inline and it seems this is what @hashhar did inline

**Comment by findepi** [comment]
i was wrong https://github.com/trinodb/trino/pull/7474#issuecomment-816547700, but you did the right thing. Please squash anyway.

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
did CI hit https://github.com/trinodb/trino/issues/7534 ?

**Comment by hashhar** [comment]
> did CI hit #7534 ?

Yes. It seems it did hit https://github.com/trinodb/trino/issues/7534.
Seems to be a racy random port allocation (I think it's not actually random because this seems to fail quite a lot).

**Comment by findepi** [comment]
> Seems to be a racy random port allocation (I think it's not actually random because this seems to fail quite a lot).

let's talk this under that issue

**Comment by findepi** [comment]
see other CI red

**Comment by hashhar** [comment]
SQL server ran out of connections. I haven't seen this before (but it's unrelated) but will keep an eye out for repeat occurences.

```
Error:  init(io.trino.plugin.sqlserver.TestSqlServerWithoutSnapshotIsolation)  Time elapsed: 151.299 s  <<< FAILURE!
java.lang.RuntimeException: Failed to execute statement: CREATE DATABASE database_0e8bf7a0b1294603b3577e98c59ed906
	at io.trino.plugin.sqlserver.TestingSqlServer.execute(TestingSqlServer.java:60)
	at io.trino.plugin.sqlserver.TestingSqlServer.setUpDatabase(TestingSqlServer.java:93)
	at io.trino.plugin.sqlserver.TestingSqlServer.start(TestingSqlServer.java:73)
	at io.trino.plugin.sqlserver.TestSqlServerWithoutSnapshotIsolation.createQueryRunner(TestSqlServerWithoutSnapshotIsolation.java:36)
	at io.trino.testing.AbstractTestQueryFramework.init(AbstractTestQueryFramework.java:93)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:169)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.sql.SQLException: Could not create new connection
	at org.testcontainers.containers.JdbcDatabaseContainer.createConnection(JdbcDatabaseContainer.java:228)
	at io.trino.plugin.sqlserver.TestingSqlServer.execute(TestingSqlServer.java:55)
	... 17 more
```

**Pull request merged**